### PR TITLE
RavenDB-19704 Sharding - Replication break for RevisionTombstoneRepli…

### PR DESCRIPTION
**Pull request by shiranshalom**
…cationItem

### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-19704/Sharding-Replication-break-for-RevisionTombstoneReplicationItem

### Type of change

- Bug fix

### How risky is the change?

- Moderate 

### Backward compatibility

- Non breaking change

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

**Comment by karmeli87** [comment]
I want to have a proper fix as well.
I was thinking that the new `Id` for the revision tombstone should be something like `DocId|revisionChangeVector`

**Comment by ayende** [comment]
How would you handle backward compact?

**Comment by karmeli87** [comment]
Probably will need to bump the replication protocol and strip the `DocId`

**Comment by shiranshalom** [comment]
> I want to have a proper fix as well. I was thinking that the new `Id` for the revision tombstone should be something like `DocId|revisionChangeVector`

Done

File: src/Raven.Server/Documents/Replication/DocumentInfoHelper.cs [hunk 1]
```diff
@@ -24,7 +24,24 @@ public unsafe LazyStringValue GetDocumentId(Slice key)
...
+            int sizeOfDocId = 0;
+            var tmpStr = key.ToCharArray();
+            for (; sizeOfDocId < key.Size; sizeOfDocId++)
+            {
+                if (tmpStr[sizeOfDocId] == (char)SpecialChars.RecordSeparator)
+                    break;
+            }
+
+            if (sizeOfDocId >= key.Size)
+                return null;
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Replication/DocumentInfoHelper.cs]: `key.IndexOf(...)`

File: src/Raven.Server/Documents/Replication/ReplicationItems/RevisionTombstoneReplicationItem.cs [hunk 2]
```diff
@@ -78,6 +80,32 @@ protected override ReplicationBatchItem CloneInternal(JsonOperationContext conte
...
+            int sizeOfDocId = 0;
+            for (; sizeOfDocId < size; sizeOfDocId++)
+            {
+                if (p[sizeOfDocId] == SpecialChars.RecordSeparator)
+                    break;
+            }
+
+            var changeVectorIndex = sizeOfDocId + 1;
+            if (changeVectorIndex >= size)
+                return;
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Replication/ReplicationItems/RevisionTombstoneReplicationItem.cs]: `Id.IndexOf(...)`

File: src/Raven.Server/Documents/Replication/ReplicationItems/RevisionTombstoneReplicationItem.cs [hunk 3]
```diff
@@ -78,6 +80,32 @@ protected override ReplicationBatchItem CloneInternal(JsonOperationContext conte
...
+
+            Id = context.AllocateStringValue(null, p + changeVectorIndex, size - sizeOfDocId - 1);
+        }
+
+        public static ByteStringContext.InternalScope TryExtractChangeVectorSliceFromKey(ByteStringContext allocator, LazyStringValue key, out Slice changeVectorSlice)
+        {
+            var index = key.IndexOf((char)SpecialChars.RecordSeparator, StringComparison.OrdinalIgnoreCase);
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Replication/ReplicationItems/RevisionTombstoneReplicationItem.cs:100]: let's do 
```
if(index == -1)
{
  ...
}
else
{
  ...
}
```

File: src/Raven.Server/Documents/Replication/Senders/MigrationReplicationDocumentSender.cs [hunk 4]
```diff
@@ -26,7 +26,8 @@ public MigrationReplicationDocumentSender(Stream stream, OutgoingMigrationReplic
             Parent = parent;
         }
 
-        protected override IEnumerable<ReplicationBatchItem> GetReplicationItems(DocumentsOperationContext ctx, long etag, ReplicationStats stats, bool caseInsensitiveCounters)
+        protected override IEnumerable<ReplicationBatchItem> GetReplicationItems(DocumentsOperationContext ctx, long etag, ReplicationStats stats, 
+            bool caseInsensitiveCounters, bool revisionTombstonesWithId)
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Replication/Senders/MigrationReplicationDocumentSender.cs]: move those `bool`s into a class

File: src/Raven.Server/Documents/Revisions/RevisionsStorage.cs [hunk 5]
```diff
@@ -694,7 +694,9 @@ public IEnumerable<string> GetCollections(Transaction transaction)
 
                         hasValue = true;
 
-                        using (Slice.From(context.Allocator, revision.ChangeVector, out var keySlice))
+                        using (Slice.From(context.Allocator, revision.ChangeVector, out var changeVectorSlice))
+                        using (context.Allocator.Allocate(prefixSlice.Size + changeVectorSlice.Size, out var keyBuffer))
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Revisions/RevisionsStorage.cs]: the prefix here is the id of the document?

Comment (karmeli87) [reply src/Raven.Server/Documents/Revisions/RevisionsStorage.cs]: let move this all inside the `CreateRevisionTombstoneKeySlice`

File: src/Raven.Server/Documents/Revisions/RevisionsStorage.cs [hunk 6]
```diff
@@ -727,13 +729,30 @@ public IEnumerable<string> GetCollections(Transaction transaction)
...
             var collectionName = _documentsStorage.ExtractCollectionName(context, collection);
             var table = EnsureRevisionTableCreated(context.Transaction.InnerTransaction, collectionName);
 
             long revisionEtag;
-            if (table.ReadByKey(key, out TableValueReader tvr))
+
+            Slice changeVectorSlice = key;
+            if (revisionTombstoneKey != null)
+            {
+                using var scope = RevisionTombstoneReplicationItem.TryExtractChangeVectorSliceFromKey(context.Allocator, revisionTombstoneKey, out changeVectorSlice);
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Revisions/RevisionsStorage.cs]: you dispose it immediately here, that can be a problem, no?

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs [hunk 7]
```diff
@@ -185,6 +184,24 @@ private ReplicationBatches PrepareReplicationDataForShards(TransactionOperationC
...
+                {
+                    if (SupportedFeatures.Replication.RevisionTombstonesWithId)
+                    {
+                        var docId = _documentInfoHelper.GetDocumentId(revisionTombstoneItem);
+                        if (docId != null)
+                        {
+                            batches[_parent.Context.GetShardNumberFor(context, docId)].Items.Add(item);
+                            continue;
+                        }
+                    }
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs:198]: this already done under `GetShardNumberForReplicationItem`, no?

File: src/Raven.Server/Documents/Replication/DocumentInfoHelper.cs (hunk 1 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Replication/DocumentInfoHelper.cs]: Done

File: src/Raven.Server/Documents/Replication/ReplicationItems/RevisionTombstoneReplicationItem.cs (hunk 2 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Replication/ReplicationItems/RevisionTombstoneReplicationItem.cs]: Done

File: src/Raven.Server/Documents/Replication/ReplicationItems/RevisionTombstoneReplicationItem.cs (hunk 3 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Replication/ReplicationItems/RevisionTombstoneReplicationItem.cs:100]: Done

File: src/Raven.Server/Documents/Replication/Senders/MigrationReplicationDocumentSender.cs (hunk 4 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Replication/Senders/MigrationReplicationDocumentSender.cs]: Done

File: src/Raven.Server/Documents/Revisions/RevisionsStorage.cs (hunk 5 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Revisions/RevisionsStorage.cs]: Done

File: src/Raven.Server/Documents/Revisions/RevisionsStorage.cs (hunk 6 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Revisions/RevisionsStorage.cs]: Done

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs (hunk 7 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs:198]: Yes, but it won't work or maybe return the wrong result if `DocId `is null so I prefer checking it directly here

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs [hunk 8]
```diff
@@ -252,10 +269,8 @@ private int GetShardNumberForReplicationItem(TransactionOperationContext context
                 case DocumentReplicationItem:
                 case TimeSeriesDeletedRangeItem:
                 case TimeSeriesReplicationItem:
-                    return _parent.Context.GetShardNumberFor(context, _documentInfoHelper.GetDocumentId(item));
                 case RevisionTombstoneReplicationItem:
-                    DevelopmentHelper.ShardingToDo(DevelopmentHelper.TeamMember.Shiran, DevelopmentHelper.Severity.Normal, "Handle this (document Id does not exist)");
-                    throw new NotSupportedInShardingException("TODO: implement for sharding"); // revision tombstones doesn't contain any info about the doc. The id here is the change-vector of the deleted revision
+                    return _parent.Context.GetShardNumberFor(context, _documentInfoHelper.GetDocumentId(item));
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs]: let's throw if `_documentInfoHelper.GetDocumentId(item)` is null

File: test/InterversionTests/ReplicationTests.cs [hunk 9]
```diff
@@ -111,6 +122,73 @@ public async Task ShouldNotReplicateIncrementalTimeSeriesToOldServer()
             }
         }
 
+        [RavenFact(RavenTestCategory.Replication | RavenTestCategory.Revisions | RavenTestCategory.Sharding)]
+        public async Task ExternalReplicationWithRevisionTombstones_ShardedToOldServer()
+        {
+            var processNode = await GetServerAsync("5.3.0-nightly-20211107-0402");
```
Comment (karmeli87) [inline test/InterversionTests/ReplicationTests.cs]: let's use here latest 5.4

File: test/InterversionTests/ReplicationTests.cs [hunk 10]
```diff
@@ -163,5 +241,276 @@ private static async Task<ModifyOngoingTaskResult> SetupReplication(IDocumentSto
...
+                //todo fix timeseries migration
+                /*var val = session.TimeSeriesFor("users/1$usa", "Heartrate")
+                    .Get(DateTime.MinValue, DateTime.MaxValue);
+
+                Assert.Equal(1, val.Length);
+
+                val = session.TimeSeriesFor("users/2$usa", "Heartrate")
+                    .Get(DateTime.MinValue, DateTime.MaxValue);
+
+                Assert.Equal(1, val.Length);*/
```
Comment (karmeli87) [inline test/InterversionTests/ReplicationTests.cs]: what is that?

File: src/Raven.Server/Documents/Revisions/RevisionsStorage.cs [hunk 11]
```diff
@@ -727,13 +727,26 @@ public IEnumerable<string> GetCollections(Transaction transaction)
             return deletedRevisionsCount;
         }
 
-        public void DeleteRevision(DocumentsOperationContext context, Slice key, string collection, string changeVector, long lastModifiedTicks)
+        private unsafe ByteStringContext.ExternalScope CreateRevisionTombstoneKeySlice(DocumentsOperationContext context, Slice documentIdSlice, string changeVector, out Slice changeVectorSlice, out Slice keySlice)
+        {
+            Slice.From(context.Allocator, changeVector, out changeVectorSlice);
+            context.Allocator.Allocate(documentIdSlice.Size + changeVectorSlice.Size, out var keyBuffer);
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Revisions/RevisionsStorage.cs]: those need to be disposed, no?

**Comment by karmeli87** [comment]
also please rebase

File: test/InterversionTests/ReplicationTests.cs (hunk 9 above)
Comment (shiranshalom) [reply test/InterversionTests/ReplicationTests.cs]: Done

File: test/InterversionTests/ReplicationTests.cs (hunk 10 above)
Comment (shiranshalom) [reply test/InterversionTests/ReplicationTests.cs]: Removed :-)

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs (hunk 8 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs]: done

File: src/Raven.Server/Documents/Revisions/RevisionsStorage.cs (hunk 11 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Revisions/RevisionsStorage.cs]: done

**Review by ppekrol — APPROVED** [review]

File: src/Raven.Server/Documents/Replication/DocumentInfoHelper.cs [hunk 12]
```diff
@@ -26,7 +26,17 @@ public unsafe LazyStringValue GetDocumentId(Slice key)
...
         }
-        
+
+        public unsafe LazyStringValue GetDocumentId(LazyStringValue key)
+        {
+            var index = key.IndexOf((char)SpecialChars.RecordSeparator, StringComparison.OrdinalIgnoreCase);
+            if (index == -1)
+                return null;
+
+            _tmpLazyStringInstance = _context.GetLazyString(key.Buffer, index);
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/DocumentInfoHelper.cs]: Why do you store that in the temporary variable?

**Review by ayende — APPROVED** [review]

File: src/Raven.Server/Documents/Replication/DocumentInfoHelper.cs (hunk 12 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Replication/DocumentInfoHelper.cs]: done

**Pull request merged**